donkeycar/parts/tub_v2.py

```diff
@@ -120,13 +124,48 @@ def run(self, *args):
         self.tub.write_record(record)
         return self.tub.manifest.current_index
 
+    def run_threaded(self, *args):
+        assert len(self.tub.inputs) == len(args), \
+            f'Expected {len(self.tub.inputs)} inputs but received {len(args)}'
+        record = dict(zip(self.tub.inputs, args))
+        # push the record into queue
+        self.record_queue.put(record)
+        self.mutex.acquire()
```

Comment:
I believe you don't need a lock here as you are just reading the counter from the tub manifest.

Reply:
I do believe you need a lock. Manifest class has nothing protect it from accessing from multiple threads. Here `Manifest.current_index` has a read access from main thread, and in `update`, which is called on part's worker thread, will in turn call `Manifest.write_record` which will change its `current_index`(write access). The read/write of a variable in Python isn't atomic. Unless I missed something.

Reply:
There is only 1 dedicated thread writing a record. So you don't need a lock.

Reply:
> There is only 1 dedicated thread writing a record. So you don't need a lock.

There's only 1 deciated thread writing the records, true. But there's another thread reading some metadata of the records. In this case, it's `run_threaded` function that needs to return the number of the records already written. So here it's a one reader and one writer problem. It needs a lock. If we can avoid returning the number of records in `run_threaded`, then we can remove this lock here.

Reply:
@tikurahul - do you have any objections left here?